manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,857 @@
+	if ipam == nil {
+		return errors.ErrInvalidSpec(
+			"ipam driver %v cannot be found", local.nw.IPAM.Driver.Name,
+		)
+	}
+	ipamOpts := local.nw.IPAM.Driver.Options
+	// NOTE(dperny): this code, where we try parsing as CIDR and
+	// then as a regular IP, is from the old allocator. I do not
+	// know why this is done this way
+	addr, _, err := net.ParseCIDR(address)
+	if err != nil {
+		addr = net.ParseIP(address)
+		if addr == nil {
+			return errors.ErrInvalidSpec("address %v is not valid", address)
+		}
+	}
+	// we don't know which pool this address belongs to, so go through each
+	// pool in this network and try to request this address.
+	//
+	// NOTE(dperny): this code is couched in 2 assumptions:
+	//   - IPAM pools can't overlap
+	//   - IPAM address requests will prefer "out of range" to "no available
+	//	   IPs"
+	// The first one I'm rather sure of, but the second i'm not... I don't know
+	// what the response would be if the pool had no remaining addresses but
+	// the requested address was out of range anyway
+	for _, poolID := range local.pools {
+		ip, _, err := ipam.RequestAddress(poolID, addr, ipamOpts)
+		if err == ipamapi.ErrIPOutOfRange {
+			continue
+		}
+		if err != nil {
+			return errors.ErrBadState(
+				"error restoring network %v when requesting address %v: %v",
+				local.nw.ID,
+				addr.String(),
+				err,
+			)
+		}
+		// if we get this far, the address belongs to this pool. add to the
+		// endpoints map for deallocation later and return nil, for no error
+		local.endpoints[ip.String()] = poolID
+		return nil
+	}
+
+	// if we get all the way through this loop, without jumping to
+	// the next iteration of the addresses loop, then we're in a
+	// weird situation where the address is out of range for
+	// _every_ pool on the network.
+	return errors.ErrBadState(
+		"error restoring network %v when requesting address %v: %v",
+		local.nw.ID,
+		addr.String(),
+		"address is out of range for all pools",
+	)
+}
+
+// AllocateNetwork allocates the IPAM pools for the given network. The network
+// must not be nil, and must not use a node-local driver.
+//
+// We can't use a node-local driver, because we don't have access to the
+// network's driver information. We only have acccess to the IPAM driver.
+//
+// If successful, this method will mutate the provided network, but will not
+// otherwise.
+func (a *allocator) AllocateNetwork(n *api.Network) (rerr error) {
+	// check if this network is already being managed and return an error if it
+	// is. networks are immutable and cannot be updated.
+	if _, ok := a.networks[n.ID]; ok {
+		return errors.ErrAlreadyAllocated()
+	}
+
+	// Now get the IPAM driver and options, either the defaults or the user's
+	// specified options.
+	ipamName := ipamapi.DefaultIPAM
+	if n.Spec.IPAM != nil && n.Spec.IPAM.Driver != nil && n.Spec.IPAM.Driver.Name != "" {
+		ipamName = n.Spec.IPAM.Driver.Name
+	}
+	ipam, _ := a.drvRegistry.IPAM(ipamName)
```

Comment:
i think not actually. we should check in the caller, and not pass node-local networks into the ipam allocator (because they don't need IPAM state allocated), which we already do.

Reply:
this is specified in the doc comment already, as well.

Reply:
Got it - just wanted to make sure we didn't need to enforce it in this function.